### PR TITLE
fixes texane/stlink#780

### PR DESCRIPTION
**Pull request by ceremcem**

### CODE REVIEW

File: src/usb.c [hunk 1]
```diff
@@ -52,7 +52,7 @@ ssize_t send_recv(struct stlink_libusb* handle, int terminate,
             3000);
     if (t) {
         printf("[!] send_recv send request failed: %s\n", libusb_error_name(t));
-        return -1;
+        exit(1);
```
Comment (xor-gate) [inline src/usb.c]: this solution is not acceptable as stlink can also be used as a library, and a library may never use exit. probably the higher layer should catch the error and exit.

* https://github.com/texane/stlink/blob/master/src/gdbserver/gdb-server.c#L264-L267
* https://github.com/texane/stlink/blob/master/src/gdbserver/gdb-server.c#L1063 
** This function complexity is insane, a lot of `stlink_` calls are not checked for error return values and the main loop should exit not the library itself

Let me know what you think, its a dirty job to fix but its worth it.

Comment (ceremcem) [reply src/usb.c]: Does every `-1` response mean `exit` in the higher layer?

Comment (xor-gate) [reply src/usb.c]: That is a very good question! IMHO when the libusb backend call fails, indeed there is some sort of failure with the I/O which most of the time you are unable to recover. E.g a stlink instance destroy is nessary, rescan of usb devices and attach again or try with exponential backoff. This is the behaviour `openocd` has when you "hotplug" an stlink programmer which is very nice.

**Comment by xor-gate** [comment]
As my comment before, I think in the upper layers we should catch the error -1 and exit (or retry with exponential backoff using rescan). I think a different implementation is necessary for that. Do you feel to fix this or leave it as is ?

File: src/usb.c (hunk 1 above)
Comment (olikraus) [reply src/usb.c]: Only my opinion: If the backend fails, then it should return an error code (which is "-1").
What I have seen so far, the higher levels will react accordingly to the -1 return value.
So: I would keep this as it is.

Comment (Nightwalker-87) [reply src/usb.c]: @xor-gate mentioned an important point here. Therefore I don't see a reason for changing this either.

**Comment by ceremcem** [comment]
Sorry for being unresponsive for that long time. I had had many distractions and I still have. Current workaround works for me but it is apparently a dirty hack and needs to be improved as you suggested. I'm willing to implement the proper solution ASAP.

**Comment by Nightwalker-87** [comment]
@ceremcem: Thx for your contribution. Could you target the develop branch for that PR? That would be great.

**Comment by ceremcem** [comment]
> Could you target the develop branch for that PR?

Sure thing. 

> Can we be sure in the meanwhile that this does not break anything currently working?

I'll take extreme caution on this.

**Comment by Nightwalker-87** [comment]
@ceremcem: Any progress on this?

File: src/usb.c (hunk 1 above)
Comment (Nightwalker-87) [reply src/usb.c]: Conversation continues on 21.02.2020.

**Comment by Nightwalker-87** [comment]
@ceremcem: This PR will be closed in 3 days, should there be no further feedback.

**Comment by ceremcem** [comment]
Hi! I'm on this one at last. 

Any fix would need a test case in order to prevent any breakage in the higher layers. What would you recommend? Should I write `my-st-util` or is there any application I can use as a test case?

**Comment by Nightwalker-87** [comment]
Can you just use `st-util` from the `develop` branch?

**Comment by Nightwalker-87** [comment]
Closed by PR #913.

**Pull request closed without merging**